File: MyLeetCode_Cpp/LeetCode/Problem_811_SubdomainVisits.cpp
// LeetCode Template
// Created by dvlcis @ 2022/10/5 12:05
// 

#include  <stdio.h>
#include  <string>
#include  <iostream>
#include  <vector>
#include  <unordered_map>
#include  <set>

using namespace std;

class TrieNode {
public:
    int sum; // 节点累加值
    unordered_map<string, TrieNode *> nexts;

    TrieNode() {
        sum = 0;
    }
};

//字符串分割函数
std::vector<std::string> split(std::string str, std::string pattern) {
    std::string::size_type pos;
    std::vector<std::string> result;
    str += pattern;//扩展字符串以方便操作
    int size = str.size();
    for (int i = 0; i < size; i++) {
        pos = str.find(pattern, i);
        if (pos < size) {
            std::string s = str.substr(i, pos - i);
            result.push_back(s);
            i = pos + pattern.size() - 1;
        }
    }
    return result;
}

class Problem_811_SubdomainVisits {
private:
    TrieNode *root = new TrieNode();

    void addNode(int count, string s) {
        TrieNode *cur = root;
        auto ss = split(s, ".");
        for (int i = ss.size() - 1; i >= 0; i--) {
            if (cur->nexts.count(ss[i]) == 0) {
                cur->nexts[ss[i]] = new TrieNode();
            }
            cur = cur->nexts[ss[i]];
            cur->sum += count;
        }
    }

public:


    void dfs(TrieNode *root, string path, vector<string>& ans) {
        if (root == nullptr) return;
        for (auto &&[subdomain, node]: root->nexts) {
            int cnt = node->sum;
            string domain;
            if (path.empty()) {
                domain = subdomain;
            } else {
                domain = subdomain + "." + path;
            }
            ans.emplace_back(to_string(cnt) + " " + domain);
            dfs(root->nexts[subdomain], domain, ans);
        }

    }

    vector<string> subdomainVisits(vector<string> &cpdomains) {
        for (auto &&s: cpdomains) {
            int spPos = s.find(' ');
            int count = stoi(s.substr(0, spPos));
            string domain = s.substr(spPos + 1);
            addNode(count, domain);
        }
        vector<string> ans;
        dfs(root, "", ans);
        return ans;
    }
};

// for test
int main() {
    Problem_811_SubdomainVisits *sl = new Problem_811_SubdomainVisits();
    vector<string> cpdomains = {"9001 discuss.leetcode.com"};
    auto ans = sl->subdomainVisits(cpdomains);
    for (auto &s: ans) {
        cout << s << " ,";
    }
    cout << endl;
    return 0;
}

File: MyLeetCode_Java/src/main/java/LeetCode/Problem_811_SubdomainVisits.java
package LeetCode;

import java.util.*;

public class Problem_811_SubdomainVisits {


    public static class TrieNode {
        public int sum;
        public boolean end;
        public Map<String, TrieNode> nexts;

        public TrieNode() {
            sum = 0;
            end = false;
            nexts = new HashMap<>();
        }
    }

    private TrieNode root = new TrieNode();

    public List<String> subdomainVisits(String[] cpdomains) {

        for (String cp : cpdomains) {
            String[] ss = cp.split(" ");
            int cnt = Integer.parseInt(ss[0]);
            String[] domains = ss[1].split("\\.");
            add(domains, cnt);
        }

        List<String> ans = query();
        return ans;
    }

    private List<String> query() {
        List<String> ans = new LinkedList<>();
        process(root, new LinkedList<>(), ans);

        return ans;
    }

    private void process(TrieNode root, LinkedList<String> path, List<String> ans) {
        if (root == null) {
            return;
        }
        Map<String, TrieNode> map = root.nexts;
        for (String next : map.keySet()) {
            int cnt = map.get(next).sum;
            // String curDomain = next;
            path.add(0, next);
            genAns(cnt, path, ans);
            process(map.get(next), path, ans);
            path.remove(0);
        }

    }

    private void genAns(int cnt, LinkedList<String> path, List<String> ans) {
        StringBuilder sb = new StringBuilder();
        sb.append(cnt).append(" ");
        for (int i = 0; i < path.size() - 1; i++) {
            sb.append(path.get(i)).append(".");
        }
        sb.append(path.get(path.size() - 1));
        ans.add(sb.toString());
    }

    private void add(String[] domains, int cnt) {
        TrieNode cur = root;
        int N = domains.length;
        for (int i = N - 1; i >= 0; i--) {
            String s = domains[i];
            if (!cur.nexts.containsKey(s)) {
                cur.nexts.put(s, new TrieNode());
            }
            cur = cur.nexts.get(s);
            cur.sum += cnt;
        }
        cur.end = true;
    }

    public static void main(String[] args) {
        // String[] cpdomains = {"9001 discuss.leetcode.com"};
        String[] cpdomains = {"900 google.mail.com", "50 yahoo.com", "1 intel.mail.com", "5 wiki.org"};
        var ans = new Problem_811_SubdomainVisits().subdomainVisits(cpdomains);
        System.out.println(ans);
    }
}
